crates/bevy_ecs/src/error/handler.rs

```
@@ -1,75 +1,150 @@
-use crate::{component::Tick, error::BevyError, resource::Resource};
-use alloc::borrow::Cow;
+use core::fmt::Display;
+use std::sync::OnceLock;
 
-/// Additional context for a failed system run.
-pub struct SystemErrorContext {
-    /// The name of the system that failed.
-    pub name: Cow<'static, str>,
+use crate::{component::Tick, error::BevyError};
+use alloc::borrow::Cow;
 
-    /// The last tick that the system was run.
-    pub last_run: Tick,
+/// Context for a [`BevyError`] to aid in debugging.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum ErrorContext {
```

Future PR: Experimenting with this context type will be very helpful in the future. For example, systems and observers could include the `SystemId`, allowing a more rich error handler to have differing behaviour based on the system that failed, but entirely up to the user.

---

FYI I think that sticking complex app-specific error handling in the global error handler is likely to be a bit of a mess. Instead, users should be trying to handle errors as close to where they're generated as possible, and we should provide them the tools required to do that.

---

I agree with Alice here complex conditional logic in the default handler will probably be an anti-pattern. We can expect users to want to have complex error handling for *their code* but much less precision is possible for engine code or ecosystem code. For code you write yourself, there's probably better ways to handle errors (like piping). What the default handler gives is consistency: you don't have to worry about one ecosystem developer choosing to panic and another choosing to silently fail without logging.

That said, we should still probably let users do it. Providing the system ID can't hurt.

---

100% agreed on source location tracking, and no objection to exposing system id etc. Seems handy for tooling! Just slap on some "please don't omni-error handling" advice in the docs :)